skops/card/_card_alternative.py

```diff
@@ -0,0 +1,361 @@
+        for key, val in kwargs.items():
+            self._add_single(key, val)
+        return self
+
+    def _add_single(self, key: str, val: Formattable | str) -> None:
+        section = self._data
+        *subsection_names, leaf_node_name = split_subsection_names(key)
+        for subsection_name in subsection_names:
+            section_maybe = section.get(subsection_name)
+
+            # there are already subsections
+            if section_maybe is not None:
+                section = section_maybe.subsections
+                continue
+
+            # no subsection, create
+            entry = Section(title=subsection_name)
+            section[subsection_name] = entry
+            section = entry.subsections
+
+        if leaf_node_name in section:
+            # entry exists, only overwrite content
+            section[leaf_node_name].content = val
+        else:
+            # entry does not exist, create a new one
+            section[leaf_node_name] = Section(title=leaf_node_name, content=val)
+
+    def _add_model(self, model) -> None:
+        model = getattr(self, "model", None)
+        if model is None:
+            return
+
+        model_str = self._strip_blank(repr(model))
+        model_repr = aRepr.repr(f"model={model_str},").strip('"').strip("'")
+        self._add_single("Model description", model_repr)
+
+    def _add_model_section(self) -> None:
+        if not self.model_diagram:
+            return
+
+        model_plot_div = re.sub(r"\n\s+", "", str(estimator_html_repr(self.model)))
+        if model_plot_div.count("sk-top-container") == 1:
+            model_plot_div = model_plot_div.replace(
+                "sk-top-container", 'sk-top-container" style="overflow: auto;'
+            )
+        self._add_single("Model Plot", model_plot_div)
+
+    def _add_hyperparams(self) -> None:
+        hyperparameter_dict = self.model.get_params(deep=True)
+        table = _clean_table(
+            tabulate(
+                list(hyperparameter_dict.items()),
+                headers=["Hyperparameter", "Value"],
+                tablefmt="github",
+            )
+        )
+        self._add_single("Model description/Training Procedure/Hyperparameters", table)
+
+    def add_plot(self, folded=False, **kwargs: str) -> "Card":
+        for plot_name, plot_path in kwargs.items():
+            section = PlotSection(alt_text=plot_name, path=plot_path, folded=folded)
+            self._add_single(plot_name, section)
+        return self
+
+    def add_table(self, folded: bool = False, **kwargs: dict["str", list[Any]]) -> Card:
+        for key, val in kwargs.items():
+            section = TableSection(table=val, folded=folded)
+            self._add_single(key, section)
+        return self
+
+    def add_metrics(self, **kwargs: int | float) -> "Card":
+        self._metrics.update(kwargs)
+        self._add_metrics(self._metrics)
+        return self
+
+    def _add_metrics(self, metrics: dict[str, float | int]) -> None:
```

Comment:
One thought I had (feel free to challenge me or tell me it's a bad idea):

If we have a lot of these helper `_add_...` sections, it might be nice to split this class into two classes, one with the core functionality methods (e.g `_strip_blank` or __repr__`) and a subclass with the default formatting/helper functions, which a user might want to tweak themselves.

That way, it would be a lot easier for a user to make their *own* `MyCard` class that has different formatting/ extra methods for their own models, similar to how they might create their own template in the old .md world.

Reply:
I think instead of that they can pass something `Formatable` to `add`, and that's much easier to implement for them, isn't it?

Reply:
> That way, it would be a lot easier for a user to make their own MyCard class that has different formatting/ extra methods for their own models, similar to how they might create their own template in the old .md world.

Let me try to understand this better. Right now, if a user would like to make small modifications to the existing templates, I think this can be easily accommodated with the added dynamism of the new class. If they want to have a new default template that they can share with others, it's not easily possible, they would have to share the code (a script) that generates this new template.

What you suggest is that it should be possible to create a subclass of `Card` that starts with the new default. I like the idea. If that requires to split the class, I'm not sure, since pre-filling the card is not implemented yet (this is what I hoped to achieve with the `make_default` class method). If splitting makes it easier for the user, then I'm all for it, I would have to see it in action.

Another issue is if we want to enable completely different output formats. Just as an example, what if a user wants to create an `rst` or `org` file instead of `md`, or another `md` flavor? Right now, the `md` format is hard-coded. If we could split the class so that the "formatting engine" is decoupled, it would be a nice gain. But maybe we'll just never support other formats and users have to use `pandoc` or similar tools to achieve this.